hrjobcontract/js/test/controllers/modal/modal-contract_test.js

```diff
@@ -0,0 +1,146 @@
+          end_reason: "1",
+          notice_amount: null,
+          notice_unit: null,
+          notice_amount_employee: null,
+          notice_unit_employee: null,
+          location: "Headquarters",
+          jobcontract_revision_id: "60"
+        },
+        hour: {},
+        pay: {},
+        leave: [],
+        health: {},
+        pension: {}
+      };
+
+      // $uibModalInstanceSpy for $uibModalInstance.opened.then()
+      $uibModalInstanceSpy = {
+        opened: {
+          then: function() {
+            jasmine.createSpy('modalInstance.opened');
+          }
+        }
+      };
+
+      // $uibModalSpy for $modal.open()
+      $uibModalSpy = {
+        open: function() {
+          return {
+            result: {
+              then: function(callback) {
+                callback("edit");
+              }
+            }
+          }
+        }
+      };
+
+      $httpBackend.whenGET(/action=validatedates&entity=HRJobDetails/).respond({ success: true });
+      $httpBackend.whenGET(/action=get&entity=HRJobContract/).respond({});
+      $httpBackend.whenPOST(/action=create&entity/).respond({});
+      $httpBackend.whenPOST(/action=replace&entity/).respond({});
+      $httpBackend.whenGET(/views.*/).respond({});
+
+      makeController();
+    }));
+
+
+    describe("Sets contract end date as appropriate", function() {
+
+      beforeEach(function() {
+        createModalSpy();
+        $scope.save();
+        $rootScope.$digest();
+      });
+
+      describe("When period_end_date is null", function() {
+
+        it("Sets contract period_end_date to '' ", function() {
+          expect($scope.entity.details.period_end_date).toBe("");
+        });
+      });
+
+      describe("When period_end_date is not null", function() {
+
+        beforeEach(function() {
+          $scope.entity.details.period_end_date = '02-03-2017';
+        });
+
+        it("Sets contract period_end_date to '02-03-2017' ", function() {
+          expect($scope.entity.details.period_end_date).toBe('02-03-2017');
+        });
+      });
+    });
+
+    function makeController() {
+      $scope = $rootScope.$new();
+      $scope.allowSave = true;
+      $scope.is_primary = '';
+
+      ctrl = $controller('ModalContractCtrl', {
+        $scope: $scope,
+        $uibModal: $uibModalSpy,
+        $uibModalInstance: $uibModalInstanceSpy,
+        ContractDetailsService: ContractDetailsService,
+        action: 'edit',
+        entity: entity,
+        content: {
+          allowSave: true
+        },
+        files: {},
+        utils: {
+          contractListLen: 1
+        }
+      });
+    }
+
+    function createModalSpy() {
+      spyOn(ContractDetailsService, "validateDates").and.callFake(function() {
+        var deferred = $q.defer();
+        deferred.resolve({
+          success: true
+        });
+
+        return deferred.promise;
+      });
+
+      spyOn($uibModalSpy, "open").and.callThrough();
+    }
+  });
+});
```

Comment:
At this point, I think the whole file needs to be re written properly, here is a structure you can follow,
```javascript
define([
  'common/angular',
  'common/angularMocks',
  'job-contract/app'
], function() {
  'use strict';

  describe('ModalContractCtrl', function() {
    var ctrl, $rootScope, $controller, $scope, $q, $httpBackend, $uibModalInstanceMock,
      $uibModalMock, entity, ContractDetailsService;

    beforeEach(module('hrjc'));

    beforeEach(inject(function(_$controller_, _$rootScope_, _$httpBackend_, _$q_, _ContractService_,
      _ContractDetailsService_) {
      $controller = _$controller_;
      $rootScope = _$rootScope_;
      $httpBackend = _$httpBackend_;
      ContractDetailsService = _ContractDetailsService_;
      $q = _$q_;
    }));

    beforeEach(function() {
      $httpBackend.whenGET(/action=validatedates&entity=HRJobDetails/).respond({ success: true });
      $httpBackend.whenGET(/action=get&entity=HRJobContract/).respond({});
      $httpBackend.whenPOST(/action=create&entity/).respond({});
      $httpBackend.whenPOST(/action=replace&entity/).respond({});
      $httpBackend.whenGET(/views.*/).respond({});
    });

    beforeEach(function() {
      mockObjects();
      makeController();
    });

    describe("Sets contract end date as appropriate", function() {
      beforeEach(function() {
        createModalSpy();
      });
      
      afterEach(function () {
        $rootScope.$digest();
      });

      describe("When period_end_date is null", function() {
        beforeEach(function() {
          $scope.entity.details.period_end_date = null;
          $scope.save();
        });

        it("Sets contract period_end_date to '' ", function() {
          expect($scope.entity.details.period_end_date).toBe("");
        });
      });

      describe("When period_end_date is not null", function() {
        var mockDate = '02-03-2017';
        beforeEach(function() {
          $scope.entity.details.period_end_date = mockDate;
          $scope.save();
        });

        it("Sets contract period_end_date to '02-03-2017' ", function() {
          expect($scope.entity.details.period_end_date).toBe(mockDate);
        });
      });
    });

    function makeController() {
      $scope = $rootScope.$new();
      $scope.allowSave = true;
      $scope.is_primary = '';

      ctrl = $controller('ModalContractCtrl', {
        $scope: $scope,
        $uibModal: $uibModalMock,
        $uibModalInstance: $uibModalInstanceMock,
        ContractDetailsService: ContractDetailsService,
        action: 'edit',
        entity: entity,
        content: {
          allowSave: true
        },
        files: {},
        utils: {
          contractListLen: 1
        }
      });
    }

    function createModalSpy() {
      spyOn(ContractDetailsService, "validateDates").and.callFake(function() {
        var deferred = $q.defer();
        deferred.resolve({
          success: true
        });

        return deferred.promise;
      });

      spyOn($uibModalMock, "open").and.callThrough();
    }

    function mockObjects() {
      entity = {
        contract: {
          id: '1',
          contact_id: "04",
          deleted: "0",
          is_current: "1",
          is_primary: "1"
        },
        details: {
          id: "60",
          position: "Test-added",
          title: "Test-added",
          funding_notes: null,
          contract_type: "Apprentice",
          period_start_date: "2017-03-28",
          period_end_date: null,
          end_reason: "1",
          notice_amount: null,
          notice_unit: null,
          notice_amount_employee: null,
          notice_unit_employee: null,
          location: "Headquarters",
          jobcontract_revision_id: "60"
        },
        hour: {},
        pay: {},
        leave: [],
        health: {},
        pension: {}
      };
      // $uibModalInstanceSpy for $uibModalInstance.opened.then()
      $uibModalInstanceMock = {
        opened: {
          then: function() {
            jasmine.createSpy('modalInstance.opened');
          }
        }
      };

      // $uibModalSpy for $modal.open()
      $uibModalMock = {
        open: function() {
          return {
            result: {
              then: function(callback) {
                callback("edit");
              }
            }
          }
        }
      };
    }
  });
});
```